### DISCUSSION TITLE
How does this package detect algebraic loops in `InterconnectedSystem`?

### DISCUSSION THREAD

**Discussion by yuu5g0g0** [python-control/python-control#948] [General]
Hello!

I'm interested in the well-posedness of interconnected nonlinear systems.

How does this package detect algebraic loops in `InterconnectedSystem`?

Mainly, I want to know the meanings of the line 1102 `cycle_count = len(self.syslist) + 1` in iosys.py.
Also, please tell me a list of references for the algorithm of the algebraic loop detection if it exists.

Thank you very much and best regards.

**Comment by murrayrm** [comment]
The method is a very crude one, as you have already discovered and as indicated in the comment above the line you referenced:

    # To allow for feedthrough terms, iterate multiple times to allow
    # feedthrough elements to propagate.  For n systems, we could need to
    # cycle through n+1 times before reaching steady state
    # TODO (later): see if there is a more efficient way to compute

The basic intuition is that if there is an algebraic loop, then there is some cycle in the system graph where evaluating the output function along the cycle (without updating the states) will eventually cause the input to the block you started with to have a different value than it had initially.  The longest possible cycle is if all of the subsystems were connected in a ring, so that is why the maximum number of evaluations (initial value of `cycle_count`) is set to the number of subsystems + 1.

I'm sure there is a much better way to do this, but I was trying get something up and running quickly.  Note that there is a check in the loop so that if none of the input values change after some iteration, you can stop checking.  In particular, if none of your subsystems have a feedthrough term, you can detect that in one evaluation of all of the inputs.

**Comment by yuu5g0g0** [comment]
Thank you for providing an answer to my question.

I see the meanings of the line 1102 `cycle_count = len(self.syslist) + 1` in iosys.py.
I have an additional question.

Why is the algebraic loop detection completed in `cycle_count` times, even in case there are some loops inside a loop?

**Reply by murrayrm** [reply]
There are two ways that you could get "loops inside a loop":

* A single layer interconnected system with $n$ subsystems having interconnections that generate an "internal" cycle: in this case there will be a loop that is shorter than $n$ and so you could, in principle, run a smaller number of iterations to detect the cycle.  But running $n$ iterations will always catch this case, so it is a conservative check.

* A multi-layer interconnected systems with $n$ subsystems, and some of the subsystems might have internal loops: in this case when you call `sys._compute_static_io` there will be a recursive call to the subsystems that are themselves interconnected systems, and you will catch the cycle in the recursion.

It's possible there is some configuration where this simple approach wouldn't work, but I think it is OK.

**Comment by yuu5g0g0** [comment]
Thank you for your response.
I apologize for the late reply.

You said this simple approach is effective in most cases but I still need an intuitive understanding.
Are there mathematical bases or proofs?

**Reply by murrayrm** [reply]
I don't have any additional intuition beyond what I already provided above ("The basic intuition is").  I am not aware of a formal mathematical basis or proof, but the basic ideas above could possibly be worked into one.

**Comment by yuu5g0g0** [comment]
Thank you for your sincere reply.
